src/main/java/com/structures/helpers/Structure.java

```diff
@@ -0,0 +1,422 @@
+import net.minecraftforge.client.ForgeHooksClient;
+import net.minecraftforge.client.MinecraftForgeClient;
+import net.minecraftforge.client.model.pipeline.LightUtil;
+
+/**
+ * Structure class, used to store, create, get structures.
+ */
+public class Structure
+{
+    /**
+     * Rotation by 90 degree.
+     */
+    private static final double  NINETY_DEGREE = 90D;
+    /**
+     * Template of the structure.
+     */
+    private Template          template;
+    private Minecraft         mc;
+    private PlacementSettings settings;
+
+    private static final int   COLOR_A    = 255;
+    private static final int   COLOR_R    = 255;
+    private static final int   COLOR_G    = 255;
+    private static final int   COLOR_B    = 255;
+
+    /**
+     * Constuctor of Structure, tries to create a new structure.
+     * @param world with world.
+     * @param structureName name of the structure (at stored location).
+     * @param settings it's settings.
+     */
+    public Structure(@Nullable World world, String structureName, PlacementSettings settings)
+    {
+        InputStream inputstream = MinecraftServer.class.getResourceAsStream("/assets/" + Constants.MOD_ID + "/schematics/" + structureName + ".nbt");
+
+        if(world == null || world.isRemote)
+        {
+            this.settings = settings;
+            this.mc = Minecraft.getMinecraft();
+        }
+        if (inputstream == null)
+        {
+            return;
+        }
+
+        try
+        {
+            this.template = readTemplateFromStream(inputstream);
+        }
+        catch (IOException e)
+        {
+            Log.getLogger().warn(String.format("Failed to load template %s", structureName), e);
+        }
+        finally
+        {
+            IOUtils.closeQuietly(inputstream);
+        }
+    }
+
+    /**
+     * reads a template from an inputstream
+     */
+    private static Template readTemplateFromStream(InputStream stream) throws IOException
+    {
+        NBTTagCompound nbttagcompound = CompressedStreamTools.readCompressed(stream);
+        Template template = new Template();
+        template.read(nbttagcompound);
+        return template;
+    }
+
+    public boolean doesExist()
+    {
+        return template == null ? false : true;
+    }
+
+    public Template.BlockInfo[] getBlockInfo()
+    {
+        Template.BlockInfo[] blockList = new Template.BlockInfo[template.blocks.size()];
+        blockList = template.blocks.toArray(blockList);
+        return blockList;
+    }
+
+    public Template.BlockInfo[] getBlockInfoWithSettings(PlacementSettings settings)
+    {
+        Template.BlockInfo[] blockList = new Template.BlockInfo[template.blocks.size()];
+        blockList = template.blocks.toArray(blockList);
+
+        for (int i = 0; i < blockList.length; i++)
+        {
+            IBlockState finalState = blockList[i].blockState.withMirror(settings.getMirror()).withRotation(settings.getRotation());
```

Comment:
![MINOR](https://raw.githubusercontent.com/SonarCommunity/sonar-github/master/images/severity-minor.png) Local variable 'finalState' could be declared final [![rule](https://raw.githubusercontent.com/SonarCommunity/sonar-github/master/images/rule.png)](http://home.kk-sc.de:9000/coding_rules#rule_key=pmd%3ALocalVariableCouldBeFinal)